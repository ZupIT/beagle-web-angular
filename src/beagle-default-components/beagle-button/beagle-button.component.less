/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

beagle-button {
    .default-theme {
        border: 1px solid currentColor;
        line-height: 34px;
        color: inherit;
        background: transparent;
        border-color:rgba(36, 42, 49, 0.7);
        box-sizing: border-box;
        position: relative;
        cursor: pointer;
        outline: none;
        display: inline-block;
        white-space: nowrap;
        text-decoration: none;
        vertical-align: baseline;
        text-align: center;
        margin: 0;
        min-width: 70px;
        line-height: 36px;
        padding: 0 16px;
        border-radius: 5px;
        overflow: visible;
        &:hover {
          background-color:  rgba(36, 42, 49, 0.1);
        }
    }
}