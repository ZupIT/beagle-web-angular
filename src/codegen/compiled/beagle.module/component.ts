/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { Type } from '@angular/core'
import {
  kebabToCamelCase,
  removeExtraIndentation,
  replaceToUnderline,
} from '../../utils/formatting'
import { getComponentAnnotations } from '../../utils/metadata'
import { remoteViewSelector } from '../../../constants'


function createViewChildString(name: string, templateName: string, angularVersion: number): string {
  const componentName = replaceToUnderline(name)

  if (angularVersion >= 8) {
    return `'${componentName}': new ViewChild('${templateName}', { static: true })`
  } else {
    return `'${componentName}': new ViewChild('${templateName}')`
  }
}

function createQueries(components: Record<string, Type<any>>, angularVersion: number) {
  const componentNames = Object.keys(components)

  return componentNames.map((name) => {
    const selector = getComponentAnnotations(components[name]).selector
    const templateName = kebabToCamelCase(selector)
    return createViewChildString(name, templateName, angularVersion)
  })
}


export function createComponentString(
    components: Record<string, Type<any>>,
    angularVersion: number,
  ) {
  const queries = createQueries(components, angularVersion)

  // todo: legacy code. Remove the input "loadParams" with v2.0.
  const componentString = `
    @Component({
      selector: '${remoteViewSelector}',
      template,
      inputs: ['loadParams', 'route', 'networkOptions', 'controllerId'],
      queries: {
        ${queries.join(',\n        ')},
      },
    })
    export class BeagleRemoteView extends AbstractBeagleRemoteView {
      constructor(
        beagleProvider: BeagleProvider,
        ngZone: NgZone,
        changeDetector: ChangeDetectorRef,
      ) {
        // @ts-ignore
        super(beagleProvider, ngZone, changeDetector)
      }
    }
  `

  return removeExtraIndentation(componentString, 4)
}
