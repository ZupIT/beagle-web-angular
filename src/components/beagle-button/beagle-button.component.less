/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

@import "../assets/default-style.less";

beagle-button {
  background-color: inherit !important;
  padding: 0 !important;
  display: inline-block !important;
  
  button {
    cursor: pointer;
    border: 1px solid @swampLight;
    color: inherit;
    background: transparent;
    box-sizing: border-box;
    position: relative;
    outline: none;
    display: inline-block;
    white-space: nowrap;
    text-decoration: none;
    vertical-align: baseline;
    text-align: center;
    margin: 0;
    min-width: 70px;
    width: 100%;
    line-height: 34px;
    padding: 0 16px;
    border-radius: 5px;
    overflow: visible;

    &:hover {
      background-color: @swampTransparent;
    }
  }
}