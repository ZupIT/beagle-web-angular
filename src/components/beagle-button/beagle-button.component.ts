/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/
import { AfterViewInit, Component, Input, ViewEncapsulation } from '@angular/core'
import { ClickEvent, IdentifiableBeagleUIElement } from '@zup-it/beagle-web'
import { Properties as CSSProperties } from 'csstype'

import { BeagleComponent } from '../../runtime/BeagleComponent'
import { BeagleButtonInterface, StylesNotToInherit } from '../schemas/button'

@Component({
  selector: 'beagle-button',
  templateUrl: './beagle-button.component.html',
  styleUrls: ['./beagle-button.component.less'],
  encapsulation: ViewEncapsulation.None,
})
export class BeagleButtonComponent extends BeagleComponent
  implements BeagleButtonInterface, AfterViewInit {

  @Input() text: string
  @Input() styleId?: string
  @Input() onPress?: () => void
  @Input() style?: CSSProperties
  @Input() clickAnalyticsEvent?: ClickEvent
  @Input() enabled?: boolean
  public usefulStyle: Record<string, any> = {}
  public type = 'button'
  public disabled: boolean

  ngOnInit() {
    if (this.style) {
      this.usefulStyle = Object.keys(this.style).reduce((styleObject, prop) => {
        if (!StylesNotToInherit.includes(prop)) {
          styleObject[prop] = this.style && this.style[prop]
        }
        return styleObject
      }, {})
    }
  }

  ngOnChanges() {
    this.disabled = this.enabled === undefined ? false : !this.enabled
  }

  ngAfterViewInit() {
    if (this.getViewContentManager instanceof Function) {
      const element = this.getViewContentManager().getElement()
      this.type = this.isSubmitButton(element)
    }
  }

  handleClick() {
    this.onPress && this.type === 'button' && this.onPress()

    if (this.getViewContentManager instanceof Function) {
        const beagleService = this.getViewContentManager().getView().getBeagleService()
        const analytics = beagleService ? beagleService.analytics : null

        if (this.clickAnalyticsEvent && analytics) {
          analytics.trackEventOnClick(this.clickAnalyticsEvent)
        }
    }    
  }

  private isSubmitButton(element: IdentifiableBeagleUIElement<any> | null) {
    return element &&
      element.onPress &&
      element.onPress._beagleAction_ === 'beagle:submitForm' ? 'submit' : 'button'
  }
}
