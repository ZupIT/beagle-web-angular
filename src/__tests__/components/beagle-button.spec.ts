/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { TestBed, async, ComponentFixture } from '@angular/core/testing'
import { Properties as CSSProperties } from 'csstype'
import { BeagleButtonComponent } from '../../components/beagle-button/beagle-button.component'
import { buildBeagleTestModuleMetadata } from './test-module-builder'
import mockViewContentManager from './mocks/test-mocks.spec'

let component: BeagleButtonComponent
let fixture: ComponentFixture<BeagleButtonComponent>
const mockStyle: CSSProperties = {
  height: '100',
  padding: '10',
}

describe('BeagleButtonComponent', () => {
  beforeEach(async(() => {
    jest.autoMockOn()
    TestBed.configureTestingModule(buildBeagleTestModuleMetadata([BeagleButtonComponent]))
      .compileComponents()

    fixture = TestBed.createComponent(BeagleButtonComponent)
    component = fixture.componentInstance
    component.getViewContentManager = () => mockViewContentManager
    fixture.detectChanges()
  }))

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot()
  })

  it('should keep styles', () => {
    spyOn(component, 'ngOnInit').and.callThrough()
    component.style = mockStyle
    component.ngOnInit()
    expect(component.usefulStyle).toEqual(mockStyle)
  })

  it('should keep original styles and remove margin style', () => {
    spyOn(component, 'ngOnInit').and.callThrough()
    component.style = mockStyle
    component.ngOnInit()
    component.style = { ...mockStyle, padding: '10px' }
    expect(component.usefulStyle).toEqual(mockStyle)
  })

  it('should trigger handleClick function', () => {
    component.onPress = () => null
    fixture.detectChanges()
    spyOn(component, 'onPress')
    const button = fixture.debugElement.nativeElement.querySelector('button')
    button.click()
    expect(component.onPress).toHaveBeenCalled()
  })
})
