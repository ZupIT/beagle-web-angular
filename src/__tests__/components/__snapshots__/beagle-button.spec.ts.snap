// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeagleButtonComponent should match snapshot 1`] = `
<beagle-button
  getViewContentManager={[Function Function]}
  type={[Function String]}
  usefulStyle={[Function Object]}
>
  <button
    type="button"
  >
     
  
  </button>
</beagle-button>
`;
